core/src/main/scala/io/projectglow/vcf/VCFStreamWriter.scala

```diff
@@ -65,20 +63,62 @@ class VCFStreamWriter(
     .setOption(Options.WRITE_FULL_FORMAT_FIELD)
     .build
 
-  def write(vc: VariantContext): Unit = {
-    if (!headerHasBeenSetOrWritten) {
-      maybeSetHeaderSampleIds(vc)
-      if (writeHeader) {
-        writer.writeHeader(header)
+  def setHeader(vcBuilder: VariantContextBuilder): Unit = {
+    val samplesIdsFromMissing = if (!inferSampleIds) {
+      sampleIdsFromMissingOpt.get
+    } else {
+      val vcSamples = vcBuilder.getGenotypes.asScala.map(_.getSampleName)
+      val numTotalSamples = vcSamples.length
+      val numPresentSamples = vcSamples.count(!_.isEmpty)
+
+      if (numPresentSamples > 0) {
+        if (numPresentSamples < numTotalSamples) {
+          VCFWriterUtils.throwMixedSamplesFailure()
+        }
+        SampleIdsFromMissing.presentSamples(vcSamples)
       } else {
-        writer.setHeader(header)
+        SampleIdsFromMissing.missingSamples(numTotalSamples)
       }
-      headerHasBeenSetOrWritten = true
     }
+    val sampleIds = samplesIdsFromMissing.sampleIds.asJava
+    headerSampleSet = new JHashSet(sampleIds)
+    header = new VCFHeader(headerLineSet.asJava, sampleIds)
+    replaceSampleIds = samplesIdsFromMissing.fromMissing
+  }
 
-    val maybeReplacedVC = maybeReplaceDefaultSampleIds(vc)
-    val vcBuilder = new VariantContextBuilder(maybeReplacedVC)
-    val iterator = maybeReplacedVC.getAttributes.entrySet().iterator()
+  // Replace genotypes' missing sample IDs with those from the header
+  def replaceMissingSampleIds(vcBuilder: VariantContextBuilder): VariantContextBuilder = {
+    val oldGts = vcBuilder.getGenotypes
+    if (inferSampleIds && oldGts.size != header.getNGenotypeSamples) {
+      VCFWriterUtils.throwSampleInferenceFailure()
+    }
+    val newGts = new JArrayList[Genotype](oldGts.size)
+    var i = 0
+    while (i < oldGts.size) {
+      val oldGt = oldGts.get(i)
+      if (inferSampleIds && !oldGt.getSampleName.isEmpty) {
```

Comment:
Might be more understandable to do this check in `checkInferredSampleIds`. Here we just replace all non-empty sample ids.

Reply:
By non-empty do you mean empty?

Reply:
I also think it's more understandable to catch this upfront, before we try replacing any non-missing sample IDs.